### PR TITLE
Import export

### LINKED ISSUE #905: Ability to import & export courses
Would be very useful if courses built in the tool could be exported out, and imported into another instance.

### PR DESCRIPTION
**Pull request by dancgray**
This PR contains basic export and import functionality from authoring tool to authoring tool. The export format it specific to the authoring tool and is not designed to be used for any other purpose.

## Export

1. Exports content JSON including extensions
2. Exports course assets and asset tags
3. Exports plugins added through plugin manager 
4. Export button on the project settings cog on the dashboard:
screen shot 2017-05-09 at 14 46 29

## Import

1. Imports course structure, content and settings
2. Installs new plugins and adds to imported course. Already installed plugins are not upgraded even if the import contains a newer version.
3. Assets, asset tags and meta data are imported.
4. Import can be found in the global menu:
screen shot 2017-05-09 at 14 48 50

## Notes
Much of the initial work was done by @taylortom in branch issue/905. The helpers file in plugins/output is in anticipation of framework import #1387.

The export button location could be improved as could the naming for download, source export and the AT export. I think it all becomes a little confusing for a user.

### Fixes

#905

### CODE REVIEW

File: plugins/output/adapt/helpers.js [hunk 1]
```diff
@@ -0,0 +1,254 @@
+// LICENCE https://github.com/adaptlearning/adapt_authoring/blob/master/LICENSE
+var _ = require('underscore');
+var origin = require('../../../')();
+var yauzl = require("yauzl");
```
Comment (lc-thomasberger) [inline plugins/output/adapt/helpers.js]: missing in package.json dependencies?

File: plugins/output/adapt/import.js [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+var async = require('async');
+var fs = require("fs-extra");
+var path = require("path");
+var IncomingForm = require('formidable').IncomingForm;
+
+var origin = require('../../../')();
+var database = require("../../../lib/database");
+var logger = require("../../../lib/logger");
+var usermanager = require('../../../lib/usermanager');
+var glob = require('glob');
```
Comment (lc-thomasberger) [inline plugins/output/adapt/import.js]: missing in package.json dependencies?

**Comment by dancgray** [comment]
Thanks for spotting @lc-thomasberger, package.json updated.

**Comment by itsthisjustin** [comment]
This will be massively useful when completed. Let me know the best way I can contribute to it.

**Comment by dancgray** [comment]
@itsthisjustin, would be very happy if you could test and feedback any issues, or just review the code. 

Feel free to comment on this PR or in the [gitter channel](https://gitter.im/adaptlearning/adapt_authoring).

**Comment by harlows** [comment]
@dancgray I've been testing this branch by moving simple courses between two authoring tool installs. Everything works as described for a course containing only standard contrib plugins or when a custom plugin is installed on both ATs (and the course has assets).

There is an issue when a course contains no assets. In this (admittedly artificially simplistic) scenario the plugin directory in the export zip fails to populate with any custom plugins and the import fails. Adding assets fixes this.

The other issue I had was when the plugin was new to the importing AT. The plugin imported and appeared correctly in the plugin manager list, but the course would not preview. This seemed to be because the plugin folder structure created in the temp/[tenant_id]/adapt_framework/src/components/ directory was empty. Uploading the plugin via the plugin manager, although it failed, populated this folder and the course could be then previewed.

I hope this is helpful.

**Comment by dancgray** [comment]
Thanks for testing @harlows, it's very helpful. Will fix those issues shortly.

**Comment by taylortom** [comment]
Dang, looks like I fixed the wrong import...

**Comment by dancgray** [comment]
Closing as functionality will be refactored into a single framework import.

**Pull request closed without merging**